### PR TITLE
[CLI-193] Allow updating kafka topic partition count

### PR DESCRIPTION
**Pull request by MuweiHe**
Release Notes
-------------

Breaking Changes
- PLACEHOLDER

New Features
- Update "num.partitions" for a Kafka topic with `confluent kafka topic update --config`

Bug Fixes
- PLACEHOLDER

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----

allow updating kafka topic partition count.
when getting topic partition count for describe, call `get` for topic data directly, instead of calculating the length of partitions response.

References
----------

https://docs.confluent.io/cloud/current/api.html#tag/Topic-(v3)/operation/getKafkaTopic
https://docs.confluent.io/cloud/current/api.html#tag/Topic-(v3)/operation/updatePartitionCountKafkaTopic

Test & Review
-------------

```
mhe@C02DV0KVMD6T cli % confluent kafka topic update test --config num.partitions=26
Updated the following configuration values for topic "test":
       Name      | Value | Read-Only  
-----------------+-------+------------
  num.partitions |    26 | false      

mhe@C02DV0KVMD6T cli % confluent kafka topic update test --config segment.jitter.ms=100
Updated the following configuration values for topic "test" (read-only configs were not updated):
        Name        | Value | Read-Only  
--------------------+-------+------------
  segment.jitter.ms |     0 | true       
```

Open Questions / Follow-ups
---------------------------

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,7 +6,6 @@ require (
 	github.com/antihax/optional v1.0.0
 	github.com/aws/aws-sdk-go v1.44.266
 	github.com/billgraziano/dpapi v0.4.0
-	github.com/brianstrauch/cobra-shell v0.4.0
```
Comment (brianstrauch) [inline go.mod:9]: `go mod tidy`?

File: internal/cmd/kafka/command_topic.go [hunk 2]
```diff
@@ -134,15 +134,15 @@ func (c *authenticatedTopicCommand) getNumPartitions(topicName string) (int, err
 		return 0, err
 	}
 
-	partitionsResp, httpResp, err := kafkaREST.CloudClient.ListKafkaPartitions(kafkaClusterConfig.ID, topicName)
+	topicResp, httpResp, err := kafkaREST.CloudClient.GetKafkaTopic(kafkaClusterConfig.ID, topicName)
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_topic.go:137]:
```suggestion
	topic, httpResp, err := kafkaREST.CloudClient.GetKafkaTopic(kafkaClusterConfig.ID, topicName)
```

File: internal/cmd/kafka/command_topic_update.go [hunk 3]
```diff
@@ -6,8 +6,7 @@ import (
 
 	"github.com/spf13/cobra"
 
-	kafkarestv3 "github.com/confluentinc/ccloud-sdk-go-v2/kafkarest/v3"
-
+	v3 "github.com/confluentinc/ccloud-sdk-go-v2/kafkarest/v3"
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_topic_update.go:9]:
```suggestion
	kafkarestv3 "github.com/confluentinc/ccloud-sdk-go-v2/kafkarest/v3"
```

File: internal/cmd/kafka/command_topic_update.go [hunk 4]
```diff
@@ -109,6 +107,23 @@ func (c *authenticatedTopicCommand) update(cmd *cobra.Command, args []string) er
 		return nil
 	}
 
+	readOnlyConfigs := types.NewSet()
+	configsValues := make(map[string]string)
+
+	if hasNumPartitionsChanged {
+		updateNumPartitionsInt, err := strconv.Atoi(updateNumPartitions)
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_topic_update.go:114]: https://pkg.go.dev/strconv#ParseInt with 32 as the last argument is more correct

File: internal/cmd/kafka/command_topic.go [hunk 5]
```diff
@@ -134,15 +134,15 @@ func (c *authenticatedTopicCommand) getNumPartitions(topicName string) (int, err
...
 	if err != nil {
 		if restErr, parseErr := kafkarest.ParseOpenAPIErrorCloud(err); parseErr == nil && restErr.Code == ccloudv2.UnknownTopicOrPartitionErrorCode {
 			return 0, fmt.Errorf(errors.UnknownTopicErrorMsg, topicName)
 		}
 		return 0, kafkarest.NewError(kafkaREST.CloudClient.GetUrl(), err, httpResp)
 	}
 
-	return len(partitionsResp.Data), nil
+	return int(topicResp.PartitionsCount), nil
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_topic.go:145]: Do we need to convert this to an `int`? It looks like all of the APIs using it want an `int32`.

File: test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden [hunk 6]
```diff
@@ -0,0 +1,4 @@
+Updated the following configuration values for topic "topic-exist-rest" (read-only configs were not updated):
```
Comment (brianstrauch) [inline test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden:1]: Any way we can omit the message in parenthesis if there are no read-only configs?

**Review by brianstrauch — CHANGES_REQUESTED** [review]

File: test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden (hunk 6 above)
Comment (MuweiHe) [reply test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden:1]: I can use two different msg's?

Comment (brianstrauch) [reply test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden:1]: Or just append the part in the parenthesis to the message if you detect one or more read-only configuration values.

File: internal/pkg/errors/strings.go [hunk 7]
```diff
@@ -39,12 +39,13 @@ const (
...
-	StartingConsumerMsg      = "Starting Kafka Consumer. Use Ctrl-C to exit."
-	UpdateTopicConfigMsg     = "Updated the following configuration values for topic \"%s\":\n"
-	UpdateTopicConfigRestMsg = "Updated the following configuration values for topic \"%s\" (read-only configs were not updated):\n"
-	OmitTopicCountMsg        = "The topic count will be omitted as Kafka topics for this cluster could not be retrieved: %v"
+	StartingProducerMsg         = "Starting Kafka Producer. Use Ctrl-C or Ctrl-D to exit."
+	StoppingConsumerMsg         = "Stopping Consumer."
+	StartingConsumerMsg         = "Starting Kafka Consumer. Use Ctrl-C to exit."
+	UpdateTopicConfigMsg        = "Updated the following configuration values for topic \"%s\":\n"
+	UpdateTopicConfigRestMsg    = "Updated the following configuration values for topic \"%s\"%s:\n"
+	ReadOnlyConfigNotUpdatedMsg = "(read-only configs were not updated)"
```
Comment (brianstrauch) [inline internal/pkg/errors/strings.go:47]: Some of these variables might make more sense to be moved to `internal/cmd/kafka` if they're only used there.

Comment (brianstrauch) [reply internal/pkg/errors/strings.go:47]: (No reason for them to be public)

File: test/fixtures/output/kafka/topic/update-success-rest.golden [hunk 8]
```diff
@@ -1,4 +1,4 @@
-Updated the following configuration values for topic "topic-exist-rest" (read-only configs were not updated):
+Updated the following configuration values for topic "topic-exist-rest":
```
Comment (brianstrauch) [inline test/fixtures/output/kafka/topic/update-success-rest.golden:1]: Can you add an integration test that _does_ include the parenthesis?

File: test/fixtures/output/kafka/topic/update-success-rest.golden [hunk 9]
```diff
@@ -1,4 +1,4 @@
...
         Name       | Value | Read-Only  
 -------------------+-------+------------
   compression.type | gzip  | false
```
Comment (brianstrauch) [inline test/fixtures/output/kafka/topic/update-success-rest.golden:4]: I'm considering hiding entire columns in the human output if they're marked as `omitempty` and every row contains a default value. What do you think?

Comment (MuweiHe) [reply test/fixtures/output/kafka/topic/update-success-rest.golden]: i think it still make sense for users to see in the output, which / those changes were ineffective

File: test/fixtures/output/kafka/topic/update-success-rest.golden [hunk 10]
```diff
@@ -1,5 +1,5 @@
-Updated the following configuration values for topic "topic-exist-rest":
-        Name       | Value | Read-Only  
--------------------+-------+------------
-  compression.type | gzip  | false      
-  retention.ms     |     1 | false      
+Updated the following configuration values for topic "topic-exist-rest" (read-only configs were not updated):
+        Name       |  Value   | Read-Only  
+-------------------+----------+------------
+  compression.type | producer | true
```
Comment (brianstrauch) [inline test/fixtures/output/kafka/topic/update-success-rest.golden:4]: Should this say "gzip" or something similar?

Comment (MuweiHe) [reply test/fixtures/output/kafka/topic/update-success-rest.golden:4]: the tests are updating it to `gzip`. this config is read-only so the backend should return something different to show that it was not updated

Comment (MuweiHe) [reply test/fixtures/output/kafka/topic/update-success-rest.golden:4]: it actually is! 🔢

File: internal/cmd/kafka/command_topic_describe.go [hunk 11]
```diff
@@ -55,9 +58,12 @@ func (c *authenticatedTopicCommand) describe(cmd *cobra.Command, args []string)
 		return err
 	}
 
-	numPartitions, err := c.getNumPartitions(topicName)
+	topicData, httpResp, err := kafkaREST.CloudClient.GetKafkaTopic(kafkaClusterConfig.ID, topicName)
```
Comment (brianstrauch) [inline internal/cmd/kafka/command_topic_describe.go:61]:
```suggestion
	topic, httpResp, err := kafkaREST.CloudClient.GetKafkaTopic(kafkaClusterConfig.ID, topicName)
```

Comment (MuweiHe) [reply internal/cmd/kafka/command_topic_describe.go:61]: `kafkarestv3.TopicData` is its type too

Comment (MuweiHe) [reply internal/cmd/kafka/command_topic_describe.go:61]: I'm ok either way. wdyt？

Comment (brianstrauch) [reply internal/cmd/kafka/command_topic_describe.go:61]: I think it's easier to read this way. Up to you though!

File: test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden (hunk 6 above)
Comment (sgagniere) [reply test/fixtures/output/kafka/topic/update-success-rest-partitions-count.golden]: This is a good change. I've definitely seen at least one user that got confused by this message showing up when they didn't try to change read-only configs.

**Comment by sgagniere** [comment]
@Muwei Looking through the backlog just now, I noticed that this PR probably resolves this ticket: https://confluentinc.atlassian.net/browse/CLI-193

File: internal/pkg/ccloudv2/kafkarest.go [hunk 12]
```diff
@@ -178,3 +178,13 @@ func (c *KafkaRestClient) DeleteKafkaTopic(clusterId, topicName string) (*http.R
 func (c *KafkaRestClient) ListKafkaTopics(clusterId string) (kafkarestv3.TopicDataList, *http.Response, error) {
 	return c.TopicV3Api.ListKafkaTopics(c.context(), clusterId).Execute()
 }
+
+func (c *KafkaRestClient) UpdateKafkaTopicPartitionCount(clusterId, topicName string, newPartitionCount int32) (kafkarestv3.TopicData, error) {
+	updatePartitionCountRequestData := *kafkarestv3.NewUpdatePartitionCountRequestData(newPartitionCount)
+	resp, httpResp, err := c.TopicV3Api.UpdatePartitionCountKafkaTopic(c.context(), clusterId, topicName).UpdatePartitionCountRequestData(updatePartitionCountRequestData).Execute()
+	return resp, errors.CatchCCloudV2Error(err, httpResp)
```
[the remark this answers is not in the document]
Comment (brianstrauch) [reply internal/pkg/ccloudv2/kafkarest.go]: Fixed!

**Review by brianstrauch — APPROVED** [review]

**Pull request merged**